Mathlib/RingTheory/HahnSeries.lean

```diff
@@ -618,22 +622,111 @@ theorem order_one [MulZeroOneClass R] : order (1 : HahnSeries Γ R) = 0 := by
   · exact order_single one_ne_zero
 #align hahn_series.order_one HahnSeries.order_one
 
-instance [NonUnitalNonAssocSemiring R] : Mul (HahnSeries Γ R) where
-  mul x y :=
-    { coeff := fun a =>
-        ∑ ij in addAntidiagonal x.isPWO_support y.isPWO_support a, x.coeff ij.fst * y.coeff ij.snd
-      isPWO_support' :=
+end HahnSeries
+
+/-- We introduce a type alias for HahnSeries in order to work with scalar multiplication by series.
+If we wrote a `SMul (HahnSeries Γ R) (HahnSeries Γ V)` instance, then when `V = HahnSeries Γ R`, we
+would have two different actions of `HahnSeries Γ R` on `HahnSeries Γ V`.  See
+`Mathlib.Data.Polynomial.Module` for more discussion on this problem. -/
+@[nolint unusedArguments]
+def HahnModule (Γ R V : Type*) [PartialOrder Γ] [Zero V] [SMul R V] :=
+  HahnSeries Γ V
+
+namespace HahnModule
+
+@[ext]
+theorem ext {Γ R V} [PartialOrder Γ] [Zero V] [SMul R V]
+    (x y : HahnModule Γ R V) (coeff : x.coeff = y.coeff) : x = y := by
+  simp_all only [HahnSeries.coeff_inj]
+
+variable {V : Type*} [AddCommMonoid V] [SMul R V]
+
+instance instZero : Zero (HahnModule Γ R V) :=
+  ⟨{  coeff := 0
+      isPWO_support' := by simp }⟩
+
+instance instAddHahnModule : Add (HahnModule Γ R V) where
+  add x y :=
+    { coeff := x.coeff + y.coeff
+      isPWO_support' := (x.isPWO_support.union y.isPWO_support).mono (Function.support_add _ _) }
+
+instance instAddCommMonoidHahnMonoid : AddCommMonoid (HahnModule Γ R V) where
+  zero := 0
+  add := (· + ·)
+  add_assoc x y z := by
+    ext
+    apply add_assoc
+  zero_add x := by
+    ext
+    apply zero_add
+  add_zero x := by
+    ext
+    apply add_zero
+  add_comm := fun x y => by
+    ext
+    apply add_comm
```

Comment:
You should be able to just copy across the other instance rather than building it from scratch

bors r-

bors d+

Reply:
That looks great!  Thanks for fixing it!